internal/venice-common/src/main/resources/avro/KafkaMessageEnvelope/v11/KafkaMessageEnvelope.avsc

```diff
@@ -0,0 +1,351 @@
+                  "type": "record",
+                  "fields": [
+                    {
+                      "name": "version",
+                      "doc": "The version of current incremental push. Each incremental push is associated with a version. Both 'StartOfIncrementalPush' control message and 'EndOfIncrementalPush' contain version info so they can be paired to each other.",
+                      "type": "string"
+                    }
+                  ]
+                }, {
+                  "name": "EndOfIncrementalPush",
+                  "doc": "This ControlMessage is sent per partition by each offline incremental push job, once per partition, at the end of a incremental push",
+                  "type": "record",
+                  "fields": [
+                    {
+                      "name": "version",
+                      "doc": "The version of current incremental push. Each incremental push is associated with a version. Both 'StartOfIncrementalPush' control message and 'EndOfIncrementalPush' contain version info so they can be paired to each other.",
+                      "type": "string"
+                    }
+                  ]
+                }, {
+                  "name": "TopicSwitch",
+                  "doc": "This ControlMessage is sent by the Controller, once per partition; it will only be used in leader/follower state transition model; this control message will indicate the leader to switch to a new source topic and start consuming from offset with a specific timestamp.",
+                  "type": "record",
+                  "fields": [
+                    {
+                      "name": "sourceKafkaServers",
+                      "doc": "A list of Kafka bootstrap servers URLs where the new source topic exists; currently there will be only one URL in the list, but the list opens up the possibility for leader to consume from different fabrics in active-active replication mode.",
+                      "type": {
+                        "type": "array",
+                        "items": "string"
+                      }
+                    }, {
+                      "name": "sourceTopicName",
+                      "doc": "Name of new the source topic.",
+                      "type": "string"
+                    }, {
+                      "name": "rewindStartTimestamp",
+                      "doc": "The creation time of this control message in parent controller minus the rewind time of the corresponding store; leaders in different fabrics will get the offset of the source topic by the same start timestamp and start consuming from there; if timestamp is 0, leader will start consuming from the beginning of the source topic.",
+                      "type": "long"
+                    }
+                  ]
+                }, {
+                  "name": "ChangeCaptureTopicSwitch",
```

Comment:
I got confused by the javadoc.

So this CM will be sent by Controller to Real-time to instruct the Venice leader storage node to decide which store version will produce to changelog topic, and also this CM will be forwarded by Venice storage node to local VT, so that the follower could persist the state and resume it when becoming a new leader.

When CM is sent by Controller, it doesn't specify any offset and I assume Controller needs to send this to real-time topic in every child colo.
When Venice leader storage node processes this CM in AA mode, which means it will receive a CM per real-time topic, how will Venice Leader handle it? I saw we put an offset array in this CM, does it mean leader will send CM with the current consumed offset per real-time topic to VT? what about other CMs in other real-time topics? ignore them?
To me this CM should be maintained per colo and the consumed offset high-watermark is not required and Venice Leader storage node will send one CM per colo in AA mode.

The javadoc also mentioned something about `indicate change capture consumer client to switch to a new version topic`, what does it mean? Will Controller or Venice leader storage node will forward this CM to changelog topic as well?

Reply:
As discussed, we still need to emit the VS_TS (or version swap) control message to RT from controller. To avoid consistency issue for before image events, we need V1 and repushed VT: V2 to coordinate with pivot of RT high watermark to ensure only one leader is producing to change capture VT all the time. So the control message here still needs to carry offset high watermarks and versions to switch from and to.